File: Engine/Source/Scripting/InternalCalls/Quaternion.cpp
#include <glm/glm.hpp>
#include <mono/jit/jit.h>
#include <glm/gtx/quaternion.hpp>
#include <glm/gtx/euler_angles.hpp>
#include <Yonai/Scripting/Assembly.hpp>
#include <Yonai/Scripting/InternalCalls.hpp>

using namespace glm;

ADD_MANAGED_METHOD(Quaternion, ToEuler, void, (glm::quat* input, glm::vec3* output))
{ *output = degrees(eulerAngles(*input)); }

ADD_MANAGED_METHOD(Quaternion, FromEuler, void, (glm::vec3* input, glm::quat* output))
// { *output = toQuat(orientate3(radians(vec3(input->y, input->x, input->z)))); }
{ *output = quat(radians(*input)); }

ADD_MANAGED_METHOD(Quaternion, Multiply, void, (glm::quat* a, glm::quat* b, glm::quat* output))
{ *output = (*a) * (*b); }

ADD_MANAGED_METHOD(Quaternion, MultiplyVector3, void, (glm::quat* a, glm::vec3* b, glm::vec3* output))
{ *output = (*a) * (*b); }

ADD_MANAGED_METHOD(Quaternion, Inverse, void, (glm::quat* input, glm::quat* output))
{ *output = inverse(*input); }

ADD_MANAGED_METHOD(Quaternion, InverseInline, void, (glm::quat* input))
{ *input = inverse(*input); }

ADD_MANAGED_METHOD(Quaternion, AngleAxis, void, (float angle, glm::vec3* axis, glm::quat* output))
{ *output = glm::angleAxis(angle, *axis); }

ADD_MANAGED_METHOD(Quaternion, Normalise, void, (glm::quat* value))
{ glm::normalize(*value); }

File: Apps/YonaiScriptCore/Math/Quaternion.cs
using Yonai.IO;
using Newtonsoft.Json.Linq;
using System.Runtime.CompilerServices;

namespace Yonai
{
	[System.Diagnostics.DebuggerDisplay("({x}, {y}, {z}, {w})")]
	public struct Quaternion : ISerializable
	{
		public float x, y, z, w;

		public static Quaternion Identity => new Quaternion(0, 0, 0, 1);

		public Vector3 Euler
		{
			get
			{
				Quaternion input = this;
				_ToEuler(ref input, out Vector3 output);
				return output;
			}
			set => _FromEuler(ref value, out this);
		}

		public Quaternion Normalised
		{
			get
			{
				Quaternion copy = this;
				copy.Normalise();
				return copy;
			}
		}

		public Quaternion Inverted
		{
			get
			{
				_Inverse(ref this, out Quaternion output);
				return output;
			}
		}

		public Quaternion(float x, float y, float z, float w)
		{
			this.x = x;
			this.y = y;
			this.z = z;
			this.w = w;
		}

		public static Quaternion FromEuler(Vector3 values)
		{
			if(values == Vector3.Zero)
				return Identity;

			_FromEuler(ref values, out Quaternion output);
			return output;
		}

		public static Quaternion FromEuler(float x, float y, float z) => FromEuler(new Vector3(x, y, z));

		public void Inverse() => _InverseInline(ref this);

		public void Normalise() => _Normalise(ref this);

		public static Quaternion AngleAxis(float angle, Vector3 axis)
		{
			_AngleAxis(angle, ref axis, out Quaternion output);
			return output;
		}

		public override string ToString() => $"({x}, {y}, {z}, {w})";

		public static Quaternion operator *(Quaternion a, Quaternion b)
		{
			_Multiply(ref a, ref b, out Quaternion output);
			return output;
		}

		public static Vector3 operator *(Quaternion a, Vector3 b)
		{
			_MultiplyVector3(ref a, ref b, out Vector3 output);
			return output;
		}

		public JObject OnSerialize() =>
			new JObject(
				new JProperty("x", x),
				new JProperty("y", y),
				new JProperty("z", z),
				new JProperty("w", w)
				);

		public void OnDeserialize(JObject json)
		{
			x = json["x"].Value<float>();
			y = json["y"].Value<float>();
			z = json["z"].Value<float>();
			w = json["w"].Value<float>();
		}

		#region Internal Calls
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Multiply(ref Quaternion a, ref Quaternion b, out Quaternion output);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _MultiplyVector3(ref Quaternion a, ref Vector3 b, out Vector3 output);

		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _FromEuler(ref Vector3 input, out Quaternion output);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _ToEuler(ref Quaternion input, out Vector3 output);

		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Inverse(ref Quaternion input, out Quaternion output);
		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _InverseInline(ref Quaternion input);

		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _Normalise(ref Quaternion instance);

		[MethodImpl(MethodImplOptions.InternalCall)] private static extern void _AngleAxis(float angle, ref Vector3 axis, out Quaternion output);
		#endregion

	}
}
